### ISSUE TITLE
content_by_lua_file breaks http2

### ISSUE THREAD

**Issue by bobrik** [openresty/lua-nginx-module#809]
Versions:

```
nginx version: nginx/1.11.1
built by gcc 4.9.2 (Debian 4.9.2-10)
built with OpenSSL 1.0.2h  3 May 2016
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/local/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-ipv6 --with-threads --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_realip_module --with-http_gunzip_module --with-http_gzip_static_module --with-stream --with-stream_ssl_module --add-module=ngx_devel_kit --add-module=lua-nginx-module
```

* https://github.com/simpl/ngx_devel_kit/tarball/v0.3.0
* https://github.com/openresty/lua-nginx-module/tarball/v0.10.5

With the following config:

```
server {
    listen foo.example.com:443 ssl http2;

    ssl_certificate ssl/pew.pem;
    ssl_certificate_key ssl/pew.key;

    location / {
        content_by_lua_file /etc/nginx/lua/debug.lua;
    }
}
```

I see the following error when I do `curl --http2` or open it in any browser with http2 support:

```
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* TCP_NODELAY set
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fb6d200de00)
> GET / HTTP/1.1
> Host: foo.example.com
> User-Agent: curl/7.49.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* Unexpected EOF
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, Client hello (1):
curl: (56) Unexpected EOF
```

However, when I replace location contents with `proxy_pass http://example.com` or request over http/1.1 it works as expected.

Here's what I have in `debug.lua`:

```lua
ngx.req.read_body()

ngx.header.content_type = 'text/plain';
ngx.print(ngx.req.raw_header());
ngx.print("")

if ngx.req.get_body_data() ~= nil then
  ngx.print(ngx.req.get_body_data())
end
```

**Comment by agentzh** [comment]
It's known that ngx.req.raw_header does not work in http 2 requests at all.

**Comment by bobrik** [comment]
Now that I think about it, it makes total sense. Still, it's worth mentioning in readme.

**Comment by agentzh** [comment]
@bobrik It already throws out a Lua exception. But yeah, it's worth a note in the documentation. Just committed a patch for the docs. Thanks for the suggestion.

**Comment by bobrik** [comment]
I don't see an exception in error logs for this `server` block, but I see segfaults in main error log:

```
2016/06/23 19:45:56 [alert] 7085#7085: worker process 7096 exited on signal 11
```

**Comment by agentzh** [comment]
@bobrik It was added just recently (17 days ago) in git master:

```
commit ff7e6a7a3713390f3f120e88a10dd80124452c83
Author: Yichun Zhang (agentzh) <agentzh@gmail.com>
Date:   Mon Jun 6 14:38:51 2016 -0700

    bugfix: fixed crashes in ngx.req.raw_header() for HTTP/2 requests.

    now we always throw out a Lua exception in ngx.req.raw_header() when
    being called in HTTP/2 requests.
```

**Comment by bobrik** [comment]
Ah, nice. Thank you! 👍

**Comment by kaelzhang** [comment]
So, does `ngx.req.get_headers()` either not work in http 2 requests?

**Comment by bjoe2k4** [comment]
I have no troubles using `ngx.req.get_headers()` together with http2.

**Comment by Webifi** [comment]
For my use case, content_by_lua_file was working perfectly with http2.  Unfortunately, I updated openresty to a newer version that now included commit ff7e6a7a3713390f3f120e88a10dd80124452c83, and now my content_by_lua_file scripts are broken unless I disable http2.

So, for me, "fixed crashes in ngx.req.raw_header() for HTTP/2 requests" was a breaking change.

Is there an option to not throw a lua exception for http2 and content_by_lua_file if we want things to continue to work the way thy were?
